### PR TITLE
Zone activity rules

### PR DESCRIPTION
**Pull request by pjaspers**
This brings in support for a new kind of rule, the zone_activity-rule which allows a formula to bridge zones and activities.

## Self proof reading checklist

- [x] Did I run Pronto and fixed all warnings
- [x] Make sure all naming and code will remain understandable in 6 month by someone new to the project or by you.
- [x] Did I test the right thing?
- [x] Did I test corner cases, non happy path (defensive testing)?
- [x] Check that I used the ad-hoc patterns and created files accordingly
- [x] Check code efficiency with record intensive project
- [x] Update documentation,readme accordingly

Thanks!

### CODE REVIEW

File: lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+      # Returns an array of Orbf::RulesEngine::Variable
+      def zone_activity_formula_variables
+        org_units_count = secondary_orgunits.size.to_s
+        org_units_key = Orbf::RulesEngine::ContractVariablesBuilder::ORG_UNITS_COUNT
+
+        package.all_activities_codes.each_with_object([]) do |activity_code, array|
+          package.zone_activity_rules.each do |rule|
+            rule.formulas.each do |formula|
+              values = activity_formula_values(formula, activity_code)
```
Comment (mestachs) [inline lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb]: with the comment... may be we should name them 

```ruby
ORG_UNITS_COUNT_KEY = Orbf::RulesEngine::ContractVariablesBuilder::ORG_UNITS_COUNT

def 
    ....
      rule.formulas.each do |formula|
            expanded_expression = expand_percent_values(formula, activity_code)
            expanded_expression = expand_direct_variable_reference(expanded_expression,formula, activity_code)
            array.push(build_variable(activity_code, formula, expanded_expression))
     end

# Expand %{vars}
def expand_percent_values(formula, activity_code)
        values = activity_formula_values(formula, activity_code)
        formula.expression % values
end

# Expand direct_vars
def expand_direct_variable_reference(expression,formula, activity_code)
        org_units_count = secondary_orgunits.size.to_s
        direct_variables = all_zone_referenced_formula(formula, activity_code)
        direct_variables[ORG_UNITS_COUNT_KEY] = org_units_count
        Tokenizer.replace_token_from_expression(expression, direct_variables, {})
```

File: lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      #       {
+      #         a_values: "a_for_orgunit_1, a_for_orgunit_1",
+      #         b_values: "b_for_orgunit_1, b_for_orgunit_2"
+      #       }
+      #
+      # Returns a hash
+      def activity_formula_values(formula, activity_code)
+        package.activity_rules
+          .flat_map(&:formulas)
+          .each_with_object({}) do |activity_formula, hash|
```
Comment (mestachs) [inline lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb]: would rename hash as results as you did in the  all_zone_referenced_formula

File: lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+          key = activity_formula.code + "_values"
+          next unless formula.expression.include? key
+
+          keys = secondary_orgunits.map do |secondary_orgunit|
+            r = [package.code, activity_code, activity_formula.code, secondary_orgunit.ext_id, period]
+            suffix_for_id_activity(package.code, activity_code, activity_formula.code, secondary_orgunit.ext_id, period)
+          end
+          hash[key.to_sym] = keys.join(", ")
+          hash
```
Comment (mestachs) [inline lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb]: not necessary in each_with_object

File: lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        suffix_for_activity(
+          package.code,
+          activity_code,
+          formula.code,
+          orgunit,
+          period
+        )
+      end
+
+      def exportable_variable_key(orgunit, activity_code, formula)
```
Comment (mestachs) [inline lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb]: move to VariablesBuilderSupport ? (we copy pasted it from the other builder)

File: lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+          state:                   formula.code,
+          activity_code:           activity_code,
+          orgunit_ext_id:          ref_orgunit.ext_id,
+          formula:                 formula,
+          package:                 package,
+          exportable_variable_key: exportable_variable_key(ref_orgunit, activity_code, formula)
+        )
+      end
+
+      def variable_key(orgunit, activity_code, formula)
```
Comment (mestachs) [inline lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb]: move to VariablesBuilderSupport ? (we copy pasted it from the other builder)

File: spec/lib/orbf/rules_engine/printers/graphviz_project_printer_spec.rb [hunk 6]
```diff
@@ -40,4 +40,30 @@
...
+          kind:     kind,
+          formulas: [Orbf::RulesEngine::Formula.new(
+                       "form_for_#{kind}",
+                       "quantity_act1_verified_for_1_and_201601 * 33",
+                       "comment_for_#{kind}"
+                     )]
+        )
+        package = Orbf::RulesEngine::Package.new(
+          code:       :facility,
+          kind:       :single,
```
Comment (mestachs) [inline spec/lib/orbf/rules_engine/printers/graphviz_project_printer_spec.rb]: I think we are missing an invariant enforcement in the Package in the after_init method

```ruby
zone_related_rules = package.rules.select(&:zone_related?)
raise "#{zone_related_rules.map(&:code)} rules are zone related but the package isn't zone related"  if zone_related_rules.any? && !package.zone_kind? 
```

File: lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb (hunk 1 above)
Comment (mestachs) [reply lib/orbf/rules_engine/builders/zone_activity_formula_variables_builder.rb]: may be I should start having a clearer vocabulary, what do you think about : 

- `expand` for `%{}` (and instead of values name them expansions)
- `substitute_references` (vs expand_direct_variable_reference) (and instead direct_variables name them substitutions )

or other possible name, at some point I was speaking of instantiation when
 -  the user provided expression `formula_code := xx * 2` 
 - is turned into the `package_code_formula_code_for_orgunit_and_period := xx_for_orgunit_period * 2 ` form 

we can rename the methods : 
   `activity_formula_values` => `activity_formula_expansions`
   `all_zone_referenced_formula` => `all_zone_referenced_formula_substitutions`
